### ISSUE TITLE
Requires but not After

### ISSUE THREAD

**Issue by OlafvdSpek** [systemd/systemd#11284]
I think I understand `Requires` and `After` but I'm wondering, what is the use case for a service with a unit in `Requires` but that same unit not in `After`?

Would it make sense to have `Requires` imply `After`, at least by default?

**Comment by arvidjaar** [comment]
> it breaks compatibility

Compatibility with what? Leaving aside question why `dbus.service` requires `dbus.socket` in the first place, this unit is clearly result of misunderstanding. It does not matter in practice because `dbus.socket` is started early anyway, so we never see race condition around starting these two units in real life.

Please explain why it is so important to start both units concurrently in this case and what stops working if you do not.

While there could be use case where having `Requires` without `After` is justified (I think I have seen one but cannot remember now), this example is definitely not one of them.

**Comment by evverx** [comment]
This issue looks like a question to me and it's generally recommended to use https://lists.freedesktop.org/mailman/listinfo/systemd-devel to ask questions.

@arvidjaar if you think there's something wrong with `dbus.service` could you open another issue?

**Comment by OlafvdSpek** [comment]
@evverx 
It's a feature request..

> Would it make sense to have Requires imply After, at least by default?

**Comment by evverx** [comment]
@OlafvdSpek if it's a feature request could you be more specific about why you think `Requires` should imply `After` by default?

**Comment by OlafvdSpek** [comment]
If a service requires MySQL, but MySQL is only started (milliseconds / seconds / minutes / hours) after the service, then the requirement is basically not met.
The current semantics of `Requires` seem like a contradiction.

**Comment by evverx** [comment]
@OlafvdSpek could you send a letter to the mailing list and include all the observations there plus a link to this issue? I'm pretty sure the issue will be reopened if that's enough to change the current behavior.

**Comment by OlafvdSpek** [comment]
Will do, thanks

**Comment by OlafvdSpek** [comment]
Hmm:
> The response from the remote server was:
550 5.1.1 <systemd-devel-request@lists.freedesktop.org>: Recipient address rejected: User unknown in local recipient table

**Comment by bl33pbl0p** [comment]
@arvidjaar 

The use case for me is using Requires= and Before= together, so that it behaves like Wants= on startup (in that my job always becomes runnable before the unit I require, so it completes and cannot be cancelled/invalidated by that job) but also propagate explicit stop jobs by the manager on that unit to myself (and I stop after it).

**Comment by boucman** [comment]
Requires+Before is almost the same thing as Wants+PartOf+Before 
(subtle difference in the way Restart is handled)

Whether this is abuse or not is opened to discussion...

**Comment by bl33pbl0p** [comment]
@Olafvdspek

For manager queued jobs, Requires= will for JOB_START propagate start jobs to all required units when a transaction is built. Since invalidation only works as long as our job is in the run_queue, often people recommend After= with it so that our job does not race with completion with theirs, and waits on that job to get some kind of proper dependency semantics. Now, this all makes sense, but Requires= is the only kind of dependency type that on JOB_STOP on the object unit (that is what we pulled in) propagate that to us by walking RequiredBy= in them. Hence, it is exclusive in the way it works, therefore should not be munged with After=.

systemd is a job manager at its core, I see people are already used to doing After= on their jobs, but this is not a strict requirement if you want semantics like I did.

Now with Before=, my job does not wait on the required's unit's job for completion and is dequeued for execution with *that* job waiting on us. This is similar to Wants= (although not internally as Wants= uses the fail job mode for similar effect, I think). This means that my job can never fail due to a dependency error at startup, but then if and ever the manager triggers a JOB_STOP on them (different from the unit state changes triggering things like Restart= with JOB_RESTART (this is why it is its own type and not stop+start) or BindsTo= causing stop jobs on inactive/failed on others) then it will also walk down the RequiredBy= list and iteratively propagate stop jobs to each one of those units (I don't want PartOf= because I don't want explicit JOB_RESTART propagation, I don't want BindsTo= because its semantics are a lot more strict and it propagates jobs on almost all state changes and job triggers).
